Add an admin notice about the upcoming change in PHP requirements

Konamiman opened the pull request:
### All Submissions:

* [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

The minimum required PHP version will be 7.2 as of WooCommerce 6.5. This pull request adds a dismissable admin notice visible when PHP 7.0 or 7.1 is used:

![image](https://user-images.githubusercontent.com/937723/148522801-70055bac-9800-4dc5-b8cc-9682f27b8472.png)

The notice is created (if needed) when the theme is switched or a new version of WooCommerce is installed.

**TODO:** Add the url of the announcement post in the link included in the notice.

### How to test the changes in this Pull Request:

1. Run the following in command line to simulate a new WooCommerce version installed, this will also attempt to clear some user meta that doesn't exist yet:

```
wp eval 'WC_Admin_Notices::reset_admin_notices(); delete_user_meta(get_current_user_id(), "dismissed_php72_required_in_woo_65_notice");' --user=<your WP user name or id>
```

2. Switch to PHP 7.2 and load any admin page, e.g. orders. You shouldn't see any notice.

3. Switch to PHP 7.1, repeat step 1, and load the admin page again. You should see the notice.

4. Switch to PHP 7.2, reload the admin page (DON'T repeat step 1 this time), and you should see that the notice has disappeared.

5. Switch to PHP 7.1, repeat step 1, reload the admin page, and click the "Dismiss" button. The notice shouldn't appear anymore unless you repeat step 1.

1+2 simulates a user that was already on PHP 7.2+ at the time of upgrading to the WooCommerce release that includes this pull request, nothing will happen for these users; 3+4 simulates a user that upgrades his PHP without dismissing the notice; and 3+5 simulates a user that dismisses the notice without (or before) upgrading his PHP. 

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Add - Admin notice warning about the upcoming minimum PHP version bump

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

Konamiman committed:
Add an admin notice about the upcoming change in PHP requirements

The minimum required PHP version will be 7.2 as of WooCommerce 6.5.
This adds a dismissable admin notice to PHP 7.0 and 7.1 users.

jonathansadowski committed:
Update class-wc-admin-notices.php

ObliviousHarmony commented:
@Konamiman it looks like [this functionality already exists](https://github.com/woocommerce/woocommerce/blob/c3405cf06f7ddea3aad2185dc8541955853c2575/plugins/woocommerce/includes/admin/class-wc-admin-notices.php#L410). Could we instead look at updating this notice and reset it for those who have already dismissed it, rather than create a new notice?

beaulebens commented:
Can we tweak the wording of this to make it explicit to the merchant/user that this change _will_ impact them (vs it just being some informational thing to ignore)?

Maybe change the "An upgrade to at least PHP 7.4 is recommended" part to something like "Your server is currently running an older version of PHP, so this change will impact your store. Upgrading to at least PHP 7.4 is recommended".

Konamiman commented:
@ObliviousHarmony I think these serve different purposes. The intent of the original notice was actually to inform about the _recommended_ PHP version for WooCommerce, and that's why it was implemented long before we decided about the bump; but indeed the wording seems to suggest a hard requirement.

Given that we don't do a PHP version requirement bump often, I think that a "disposable" notice with whatever wording and links we need in each case is the way to go. I think that what we can do (after Woo 6.5) is to make the original notice more clear and use it to recommend an upgrade to 7.4.

Konamiman committed:
Changes in the notice for the PHP 7.2 bump:

- Update the link to the announcement post
- Extend the notice text with "Your server is currently running an
  older version of PHP, so this change will impact your store".

Konamiman commented:
@beaulebens Text updated with your suggestion, see the updated screenshot.

peterfabian commented on plugins/woocommerce/includes/admin/class-wc-admin-notices.php:
```diff
@@ -275,6 +312,7 @@ public static function update_notice() {
 	 * @deprecated 4.6.0
 	 */
 	public static function install_notice() {
+		// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
```
The third parameter is the replacement. I think 
1. we shouldn't ignore the escaping if we're passing a translated string to it, but 
2. we shouldn't actually pass anything as that's not a replacement function, no?

Konamiman replied:
Indeed, this looks like an attempt to convey some extra information in a way that isn't really supported by the function (but no real harm done since this parameter is only used for printing error messages).

peterfabian commented on plugins/woocommerce/includes/admin/class-wc-admin-notices.php:
```diff
@@ -52,6 +52,13 @@ public static function init() {
+				self::maybe_remove_php72_required_notice();
+			},
+			20
+		);
```
Just curious, why have you decided to use a different approach than the other `add_action` calls around? Is it because you wanted to keep the methods private?

Konamiman replied:
Yes, I don't want to add public methods that will later be removed (unless we are in the `Internal` namespace, which is not the case).

peterfabian commented:
Related: https://github.com/woocommerce/woocommerce/pull/23975

peterfabian reviewed (commented):
Thanks, works fine, I just have one note regarding security of unrelated code.

ObliviousHarmony commented on plugins/woocommerce/includes/admin/class-wc-admin-notices.php:
```diff
@@ -113,6 +120,36 @@ public static function reset_admin_notices() {
+	 */
+	private static function maybe_remove_php72_required_notice() {
+		if ( version_compare( phpversion(), '7.2', '>=' ) ) {
+			self::remove_notice( 'php72_required_in_woo_65' );
```
One concern I have is that `remove_notice()` is going to call `delete_option` on every admin page request in this current implementation. Generally speaking, it looks like most notices are only removed when something changes their validity, such as right after they fix something that the notice is about. What about checking to see whether the notice is even present before trying to remove it, either in this function, or in `remove_notice`?

Konamiman replied:
That's a good point I've added that.

ObliviousHarmony reviewed (commented):
@Konamiman To be fair, given how strong the wording was, as well as the name of the notice (`wp_php_min_requirements_notice`), I do think that the original intention is not only lost but probably irrelevant. Personally, I'd still prefer we adjust the wording of the notice and reset the dismissal in a migration, but there's probably nothing wrong with a duplicate notice. I do have one performance concern with this implementation, so I've left a comment.

Konamiman committed:
Tryy to remove th PHP 7.2 bump notice only if it actually exists.

Konamiman committed:
Fix a "__" that should be "esc_html__" (triggering deprecation notice)

peterfabian reviewed (approved):
Looks good, works fine as far as I can see. There are some question marks around which actions to use to make things efficient, but this will go away in 3 versions, so let's take the safer route.
